tooltester/src/main/java/io/dockstore/tooltester/S3Client.java

```diff
@@ -0,0 +1,116 @@
+    private static final int TOOL_ID_REPOSITORY_INDEX = 3;
+    private static final int TOOL_ID_TOOLNAME_INDEX = 4;
+    private String bucketName;
+    private AmazonS3 s3;
+
+    public S3Client() {
+        TooltesterConfig tooltesterConfig = new TooltesterConfig();
+        String s3Endpoint = tooltesterConfig.getS3Endpoint();
+        bucketName = tooltesterConfig.getS3Bucket();
+        AwsClientBuilder.EndpointConfiguration endpointConfiguration = new AwsClientBuilder.EndpointConfiguration(s3Endpoint,
+                Regions.US_EAST_1.getName());
+        s3 = AmazonS3ClientBuilder.standard().withEndpointConfiguration(endpointConfiguration).build();
+    }
+
+    /**
+     * This essentially generates the filePath of the log that will be stored on S3
+     *
+     * @param toolId       The GA4GH Tool ID
+     * @param versionName  The GA4GH ToolVersion name
+     * @param testFilePath The file that was tested (Dockerfile, test.json, etc)
+     * @param runner       The runner used to test (cwltool, cromwell, etc)
+     * @param startTime    The start time in milliseconds since epoch
+     * @return S3 key (file path)
+     * @throws UnsupportedEncodingException Could not endpoint string
+     */
+    private static String generateKey(String toolId, String versionName, String testFilePath, String runner, String startTime)
+            throws UnsupportedEncodingException {
+        List<String> pathList = new ArrayList<>();
+        pathList.add(convertToolIdToPartialKey(toolId));
+        pathList.add(URLEncoder.encode(versionName, StandardCharsets.UTF_8.toString()));
+        pathList.add(
+                URLEncoder.encode(testFilePath.replaceFirst("^Build ", "").replaceFirst("^Test ", ""), StandardCharsets.UTF_8.toString()));
+        pathList.add(URLEncoder.encode(runner, StandardCharsets.UTF_8.toString()));
+        pathList.add(URLEncoder.encode(startTime + ".log", StandardCharsets.UTF_8.toString()));
+        return String.join("/", pathList);
+    }
+
+    /**
+     * Converts the toolId into a key for s3 storage.  Used by both webservice and tooltester
+     * Workflows will be in a "workflow" directory whereas tools will be in a "tool" directory
+     * repository and optional toolname or workflowname must be encoded or else looking for logs of a specific tool without toolname (quay.io/dockstore/hello_world)
+     * will return logs for the other ones with toolnames (quay.io/dockstore/hello_world/thing)
+     *
+     * @param toolId TRS tool ID
+     * @return The key for s3
+     */
+    protected static String convertToolIdToPartialKey(String toolId) throws UnsupportedEncodingException {
+        if (toolId.startsWith("#workflow")) {
+            toolId = toolId.replaceFirst("#workflow", "workflow");
+        } else {
+            toolId = "tool/" + toolId;
+        }
+        String[] split = toolId.split("/");
+        if (split.length == MAX_TOOL_ID_STRING_SEGMENTS) {
```

Comment:
Could you comment or extract to a well-named function? Otherwise you have to read the code to figure out what it's doing.

Reply:
There's already a comment that explains what it does